k9mail/src/test/java/com/fsck/k9/helper/MergeCursorTest.java

```diff
@@ -0,0 +1,450 @@
+        verify(mCursor2).close();
+    }
+
+    @Test
+    public void testCopyStringToBuffer() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+        char[] chars = {'a', 'b', 'c'};
+        CharArrayBuffer cab = new CharArrayBuffer(chars);
+        mergeCursor.copyStringToBuffer(1, cab);
+
+        verify(mCursor1).copyStringToBuffer(1, cab);
+    }
+
+    @Test
+    public void testDeactivate() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+        mergeCursor.deactivate();
+
+        verify(mCursor1).deactivate();
+        verify(mCursor2).deactivate();
+    }
+
+    @Test
+    public void testGetBlob() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+        mergeCursor.getBlob(1);
+
+        verify(mCursor1).getBlob(1);
+    }
+
+    @Test
+    public void testGetColumnCount() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        when(mCursor1.getColumnCount()).thenReturn(7);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+
+        assertEquals(7, mergeCursor.getColumnCount());
+        verify(mCursor1).getColumnCount();
+    }
+
+    @Test
+    public void testGetColumnIndex() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        when(mCursor1.getColumnIndex("abc")).thenReturn(2);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+
+        assertEquals(2, mergeCursor.getColumnIndex("abc"));
+        verify(mCursor1).getColumnIndex("abc");
+    }
+
+    @Test
+    public void testGetColumnIndexOrThrow() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        when(mCursor1.getColumnIndexOrThrow("abc")).thenReturn(2);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+
+        assertEquals(2, mergeCursor.getColumnIndexOrThrow("abc"));
+        verify(mCursor1).getColumnIndexOrThrow("abc");
+    }
+
+    @Test
+    public void testGetColumnName() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        when(mCursor1.getColumnName(2)).thenReturn("abc");
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+
+        assertEquals("abc", mergeCursor.getColumnName(2));
+        verify(mCursor1).getColumnName(2);
+    }
+
+    @Test
+    public void testGetColumnNames() {
+        reset(mCursor1);
+        reset(mCursor2);
+
+        String[] strings = new String[3];
+        strings[0] = "aaa";
+        strings[1] = "bbb";
+        strings[2] = "ccc";
+        when(mCursor1.getColumnNames()).thenReturn(strings);
+
+        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
+
+        assertEquals(strings, mergeCursor.getColumnNames());
+        verify(mCursor1).getColumnNames();
+    }
+
+    @Test
+    public void testGetCount() {
```

Comment:
These are actually two tests. One to check if the MergeCursor count is correct. And one to check if the caching works.

```java
    @Test
    public void testGetCountResult() {
        reset(mCursor1);
        reset(mCursor2);

        when(mCursor1.getCount()).thenReturn(23);
        when(mCursor2.getCount()).thenReturn(42);
        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);

        int count = mergeCursor.getCount();

        assertEquals(23 + 42, count);
    }

    @Test
    public void testGetCountCaching() {
        reset(mCursor1);
        reset(mCursor2);

        MergeCursor mergeCursor = new MergeCursor(mCursors, mComparator);
        mergeCursor.getCount();
        reset(mCursor1);
        reset(mCursor2);

        mergeCursor.getCount();

        verify(mCursor1, never()).getCount();
        verify(mCursor2, never()).getCount();
    }
```